File: decision_tree_regression.R
# Decision Tree Regression

# Importing the dataset
dataset = read.csv('crowdfunding_stats.csv', sep = ',')
#dataset = dataset[c(2,14)]

# no feature scaling - the dataset is not based on euclidean distances
# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Decision Tree Regression to the dataset
# install.packages('rpart')
library(rpart)
regressor = rpart(formula = Y ~ PKB,
                  data = dataset,
                  control = rpart.control(minsplit = 2))

# Predicting a new result with Decision Tree Regression
y_pred = predict(regressor, data.frame(PKB = 600000))

# Non-continuous model: Visualising the Decision Tree Regression results (higher resolution)
# install.packages('ggplot2')
library(ggplot2)
x_grid = seq(min(dataset$PKB), max(dataset$PKB), 1)
ggplot() +
  geom_point(aes(x = dataset$PKB, y = dataset$Y),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame(PKB = x_grid))),
            colour = 'blue') +
  ggtitle('GDP vs Crowdfunding in OECD (Test set)') +
  xlab('GDP') +
  ylab('Crowdfunding size')

# Plotting the tree
plot(regressor)
text(regressor)

File: decision_tree_regression.py
# Decision Tree Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('crowdfunding_stats.csv', sep=',')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 13].values

# Fitting Decision Tree Regression to the dataset
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state = 88)
regressor.fit(X, y)

# Predicting a new result
y_pred = regressor.predict(2500000)


# this one is non-continuous and ugly
plt.scatter(X, y, color = 'red')
plt.plot(X, regressor.predict(X), color = 'blue')
#plt.scatter(X, regressor.predict(X), color = 'blue')
plt.title('Forecast vs. reality (SVR)')
plt.xlabel('GDB')
plt.ylabel('Crowdfunding size')
plt.show()


# Visualising the Decision Tree Regression results (in ranges)
X_grid = np.arange(min(X), max(X), 100)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, regressor.predict(X_grid), color = 'blue')
plt.title('Forecast vs. reality (Decision Tree Regression)')
plt.xlabel('GDB')
plt.ylabel('Crowdfunding size')
plt.show()

